crates/bevy_pbr/src/light.rs

```diff
@@ -479,6 +590,98 @@ fn ndc_position_to_cluster(
         .clamp(UVec3::ZERO, cluster_dimensions - UVec3::ONE)
 }
 
+// Calculate an AABB for the light in view space, returns a (Vec3, Vec3) containing min and max with
+// - x and y in view space with range [-1, 1]
+// - z in world space with view orientation, with range [-inf, -0.0001] for perspective, and [1.0000, inf] for orthographic
+fn cluster_space_light_aabb(
+    is_orthographic: bool,
+    inverse_view_transform: Mat4,
+    projection_matrix: Mat4,
+    light_sphere: &Sphere,
+) -> (Vec3, Vec3) {
+    let light_aabb_view = Aabb {
+        center: (inverse_view_transform * light_sphere.center.extend(1.0)).xyz(),
+        half_extents: Vec3::splat(light_sphere.radius),
+    };
+    let (mut light_aabb_view_min, mut light_aabb_view_max) =
+        (light_aabb_view.min(), light_aabb_view.max());
+
+    if is_orthographic {
+        // constraint z to be positive - i.e. in front of the camera
+        light_aabb_view_min.z = light_aabb_view_min.z.max(1.0);
+        light_aabb_view_max.z = light_aabb_view_max.z.max(1.0);
+    } else {
+        // constraint z to be negative - i.e. in front of the camera
+        light_aabb_view_min.z = light_aabb_view_min.z.min(-0.0001);
+        light_aabb_view_max.z = light_aabb_view_max.z.min(-0.0001);
```

Comment:
```suggestion
        // Constrain view z to be negative - i.e. in front of the camera
        // When view z is >= 0.0 and we're using a perspective projection, bad things happen.
        // At view z == 0.0, ndc x,y are mathematically undefined. At view z > 0.0, i.e. behind the camera,
        // the perspective projection flips the directions of the axes. This breaks assumptions about
        // use of min/max operations as something that was to the left in view space is now returning a
        // coordinate that for view z in front of the camera would be on the right, but at view z behind the
        // camera is on the left. So, we just constrain view z to be < 0.0 and necessarily in front of the camera.
        light_aabb_view_min.z = light_aabb_view_min.z.min(-f32::EPSILON);
        light_aabb_view_max.z = light_aabb_view_max.z.min(-f32::EPSILON);
```

Reply:
This made sense after I'd understood (a tautology) but I wanted to see it in code too so I quickly hacked up this:
```rust
use bevy::{
    math::{Vec3Swizzles, Vec4Swizzles},
    prelude::*,
};

fn main() {
    let view_pos = Vec3::new(-1.0, -1.0, -1.0);
    let projection = Mat4::perspective_infinite_reverse_rh(std::f32::consts::FRAC_PI_4, 1.0, 0.1);
    let in_front_clip = projection * view_pos.extend(1.0);
    let in_front_ndc = in_front_clip.xyz() / in_front_clip.w;
    let behind_clip = projection * view_pos.xy().extend(1.0).extend(1.0);
    let behind_ndc = behind_clip.xyz() / behind_clip.w;
    dbg!(in_front_clip);
    dbg!(in_front_ndc);
    dbg!(behind_clip);
    dbg!(behind_ndc);
}
```
gives
```
[examples/3d/3d_scene.rs:13] in_front_clip = Vec4(
    -2.4142134,
    -2.4142134,
    0.1,
    1.0,
)
[examples/3d/3d_scene.rs:14] in_front_ndc = Vec3(
    -2.4142134,
    -2.4142134,
    0.1,
)
[examples/3d/3d_scene.rs:15] behind_clip = Vec4(
    -2.4142134,
    -2.4142134,
    0.1,
    -1.0,
)
[examples/3d/3d_scene.rs:16] behind_ndc = Vec3(
    2.4142134,
    2.4142134,
    -0.1,
)
```